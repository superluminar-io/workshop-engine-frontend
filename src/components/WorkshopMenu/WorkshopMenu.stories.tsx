import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { WorkshopMenu } from "./WorkshopMenu";
import { Box, MenuItem } from "@mui/material";
import EditIcon from "@mui/icons-material/Edit";

export default {
  title: "WorkshopMenu",
  component: WorkshopMenu,
  argTypes: {
    onEdit: {
      action: "edited",
    },
  },
} as ComponentMeta<typeof WorkshopMenu>;

const Template: ComponentStory<typeof WorkshopMenu> = (args) => (
  <Box
    sx={{
      width: "100vw",
      height: "100vh",
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
    }}
  >
    <WorkshopMenu {...args} />
  </Box>
);

export const Main = Template.bind({});
Main.storyName = "main";
Main.args = {
  children: (
    <MenuItem>
      <EditIcon />
      Edit
    </MenuItem>
  ),
};
