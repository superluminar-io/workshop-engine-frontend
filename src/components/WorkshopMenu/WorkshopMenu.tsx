import React, { MouseEventHandler } from "react";
import { styled } from "@mui/material/styles";
import { IconButton, Menu, MenuItemProps, MenuProps } from "@mui/material";
import MoreVertIcon from "@mui/icons-material/MoreVert";
import {
  usePopupState,
  bindTrigger,
  bindMenu,
} from "material-ui-popup-state/hooks";

const StyledWorkshopMenu = styled((props: MenuProps) => <Menu {...props} />)(
  ({ theme }) => ({
    "& .MuiPaper-root": {
      borderRadius: 6,
      minWidth: 180,
      boxShadow:
        "rgb(255, 255, 255) 0px 0px 0px 0px, rgba(0, 0, 0, 0.05) 0px 0px 0px 1px, rgba(0, 0, 0, 0.1) 0px 10px 15px -3px, rgba(0, 0, 0, 0.05) 0px 4px 6px -2px",
      "& .MuiMenuItem-root": {
        "& .MuiSvgIcon-root": {
          fontSize: 18,
          color: theme.palette.text.secondary,
          marginRight: theme.spacing(1.5),
        },
      },
    },
  })
);

export interface WorkshopMenuProps {
  id: string;
  children:
    | React.ReactElement<MenuItemProps>
    | React.ReactElement<MenuItemProps>[];
}

export const WorkshopMenu: React.FunctionComponent<WorkshopMenuProps> = ({
  id,
  children,
}) => {
  const popupState = usePopupState({
    variant: "popover",
    popupId: `workshop-menu-${id}`,
  });

  const handleMenuItemClick =
    (eventHandler?: MouseEventHandler) => (event: React.MouseEvent) => {
      popupState.close();

      if (eventHandler) {
        eventHandler(event);
      }
    };

  return (
    <>
      <IconButton aria-label="settings" {...bindTrigger(popupState)}>
        <MoreVertIcon />
      </IconButton>
      <StyledWorkshopMenu
        elevation={0}
        anchorOrigin={{
          vertical: "bottom",
          horizontal: "right",
        }}
        transformOrigin={{
          vertical: "top",
          horizontal: "right",
        }}
        {...bindMenu(popupState)}
      >
        {React.Children.map(
          children,
          (child: React.ReactElement<MenuItemProps>) => {
            return React.cloneElement(child, {
              onClick: handleMenuItemClick(child.props.onClick),
            });
          }
        )}
      </StyledWorkshopMenu>
    </>
  );
};
