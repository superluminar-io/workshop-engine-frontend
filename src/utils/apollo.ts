import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';
import { useSession } from '@clerk/nextjs';

const httpLink = createHttpLink({
  uri: process.env.NEXT_PUBLIC_APPSYNC_URL,
});

const authLink = setContext(async (_, { headers }) => {
  const { session: { getToken } } = useSession();
  const token = await getToken();
  
  return {
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : "",
    }
  }
});

export const client = new ApolloClient({
  link: authLink.concat(httpLink),
  cache: new InMemoryCache()
});
