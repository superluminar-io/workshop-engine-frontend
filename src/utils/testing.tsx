import React from "react";
import {
  render as renderTestingLibrary,
  RenderOptions,
  RenderResult,
} from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import { theme } from "../config/theme";

export const noop = () => null;

export const render = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => {
  return renderTestingLibrary(
    <ThemeProvider theme={theme}>{ui}</ThemeProvider>,
    options
  );
};
