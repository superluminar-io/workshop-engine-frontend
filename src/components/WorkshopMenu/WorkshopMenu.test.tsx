import { noop, render } from "../../utils/testing";
import { WorkshopMenu } from "./WorkshopMenu";
import { fireEvent } from "@testing-library/react";
import { MenuItem } from "@mui/material";

describe("open", () => {
  it("renders context menu", async () => {
    const { getByRole, findByText } = render(
      <WorkshopMenu id="abc">
        <MenuItem>Edit</MenuItem>
      </WorkshopMenu>
    );

    fireEvent.click(getByRole("button"));
    await findByText("Edit");

    expect(document.body).toMatchSnapshot();
  });
});

describe("close", () => {
  it("renders icon button", () => {
    render(
      <WorkshopMenu id="abc">
        <MenuItem>Edit</MenuItem>
      </WorkshopMenu>
    );
    expect(document.body).toMatchSnapshot();
  });
});
