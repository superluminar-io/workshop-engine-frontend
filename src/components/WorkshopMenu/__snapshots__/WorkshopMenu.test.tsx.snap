// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`close renders icon button 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-0::-moz-focus-inner {
  border-style: none;
}

.emotion-0.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-0 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-0:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-0:hover {
    background-color: transparent;
  }
}

.emotion-0.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
}

<body
  style=""
>
  <div>
    <button
      aria-haspopup="true"
      aria-label="settings"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium emotion-0"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-1"
        data-testid="MoreVertIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
    </button>
  </div>
</body>
`;

exports[`open renders context menu 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-0::-moz-focus-inner {
  border-style: none;
}

.emotion-0.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-0 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-0:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-0:hover {
    background-color: transparent;
  }
}

.emotion-0.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
}

.emotion-2 {
  position: fixed;
  z-index: 1300;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
}

.emotion-2 .MuiPaper-root {
  border-radius: 6px;
  min-width: 180px;
  box-shadow: rgb(255, 255, 255) 0px 0px 0px 0px,rgba(0, 0, 0, 0.05) 0px 0px 0px 1px,rgba(0, 0, 0, 0.1) 0px 10px 15px -3px,rgba(0, 0, 0, 0.05) 0px 4px 6px -2px;
}

.emotion-2 .MuiPaper-root .MuiMenuItem-root .MuiSvgIcon-root {
  font-size: 18px;
  color: rgba(0, 0, 0, 0.6);
  margin-right: 12px;
}

.emotion-3 {
  position: fixed;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
  background-color: transparent;
  -webkit-tap-highlight-color: transparent;
  z-index: -1;
}

.emotion-4 {
  background-color: #fff;
  color: rgba(0, 0, 0, 0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 12px;
  box-shadow: rgba(0, 0, 0, 0.1) 0px 0px 0px 1px;
  max-height: calc(100% - 96px);
  -webkit-overflow-scrolling: touch;
  background-color: #fff;
  color: rgba(0, 0, 0, 0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 12px;
  box-shadow: none;
  position: absolute;
  overflow-y: auto;
  overflow-x: hidden;
  min-width: 16px;
  min-height: 16px;
  max-width: calc(100% - 32px);
  max-height: calc(100% - 32px);
  outline: 0;
}

.emotion-5 {
  list-style: none;
  margin: 0;
  padding: 0;
  position: relative;
  padding-top: 8px;
  padding-bottom: 8px;
  outline: 0;
}

.emotion-6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: "Source Sans Pro",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  min-height: 48px;
  padding-top: 6px;
  padding-bottom: 6px;
  box-sizing: border-box;
  white-space: nowrap;
  padding-left: 16px;
  padding-right: 16px;
}

.emotion-6::-moz-focus-inner {
  border-style: none;
}

.emotion-6.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-6 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-6:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-6:hover {
    background-color: transparent;
  }
}

.emotion-6.Mui-selected {
  background-color: rgba(255, 59, 97, 0.08);
}

.emotion-6.Mui-selected.Mui-focusVisible {
  background-color: rgba(255, 59, 97, 0.2);
}

.emotion-6.Mui-selected:hover {
  background-color: rgba(255, 59, 97, 0.12);
}

@media (hover: none) {
  .emotion-6.Mui-selected:hover {
    background-color: rgba(255, 59, 97, 0.08);
  }
}

.emotion-6.Mui-focusVisible {
  background-color: rgba(0, 0, 0, 0.12);
}

.emotion-6.Mui-disabled {
  opacity: 0.38;
}

.emotion-6+.MuiDivider-root {
  margin-top: 8px;
  margin-bottom: 8px;
}

.emotion-6+.MuiDivider-inset {
  margin-left: 52px;
}

.emotion-6 .MuiListItemText-root {
  margin-top: 0;
  margin-bottom: 0;
}

.emotion-6 .MuiListItemText-inset {
  padding-left: 36px;
}

.emotion-6 .MuiListItemIcon-root {
  min-width: 36px;
}

@media (min-width:600px) {
  .emotion-6 {
    min-height: auto;
  }
}

<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <button
      aria-controls="workshop-menu-abc"
      aria-haspopup="true"
      aria-label="settings"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium emotion-0"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-1"
        data-testid="MoreVertIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
    </button>
  </div>
  <div
    class="MuiModal-root MuiPopover-root emotion-2"
    id="workshop-menu-abc"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiBackdrop-invisible emotion-3"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiMenu-paper MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiPopover-paper emotion-4"
      style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
      tabindex="-1"
    >
      <ul
        class="MuiList-root MuiList-padding MuiMenu-list emotion-5"
        role="menu"
        tabindex="-1"
      >
        <li
          class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root Mui-focusVisible emotion-6"
          role="menuitem"
          tabindex="0"
        >
          Edit
        </li>
      </ul>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
